common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -338,6 +340,79 @@ jbyteArray CBBToByteArray(JNIEnv* env, CBB* cbb) {
     return byteArray.release();
 }
 
+jbyteArray CryptoBufferToByteArray(JNIEnv* env, const CRYPTO_BUFFER* buf) {
+  if (CRYPTO_BUFFER_len(buf) > INT_MAX) {
+      JNI_TRACE("buffer too large");
+      Errors::jniThrowRuntimeException(env, "buffer too large");
+      return nullptr;
+  }
+
+  int length = static_cast<int>(CRYPTO_BUFFER_len(buf));
+  jbyteArray ret = env->NewByteArray(length);
+  if (ret == nullptr) {
+    JNI_TRACE("allocating byte array failed");
+    return nullptr;
+  }
+
+  env->SetByteArrayRegion(ret, 0, length, reinterpret_cast<const int8_t*>(CRYPTO_BUFFER_data(buf)));
+  return ret;
+}
+
+bssl::UniquePtr<CRYPTO_BUFFER> ByteArrayToCryptoBuffer(JNIEnv* env, const jbyteArray array,
+                                                       CRYPTO_BUFFER_POOL* pool) {
+    if (array == nullptr) {
+        JNI_TRACE("array was null");
+        Errors::jniThrowNullPointerException(env, "array == null");
+        return nullptr;
+    }
+
+    ScopedByteArrayRO arrayRo(env, array);
+    if (arrayRo.get() == nullptr) {
+        JNI_TRACE("failed to get bytes");
+        return nullptr;
+    }
+
+    bssl::UniquePtr<CRYPTO_BUFFER> ret(CRYPTO_BUFFER_new(
+            reinterpret_cast<const uint8_t*>(arrayRo.get()), arrayRo.size(), nullptr));
+    if (!ret) {
+        JNI_TRACE("failed to allocate CRYPTO_BUFFER");
+        Errors::jniThrowOutOfMemory(env, "failed to allocate CRYPTO_BUFFER");
+        return nullptr;
+    }
+
+    return ret;
+}
+
+static jobjectArray CryptoBuffersToObjectArray(JNIEnv* env,
+                                               const STACK_OF(CRYPTO_BUFFER)* buffers) {
+    size_t numBuffers = sk_CRYPTO_BUFFER_num(buffers);
+    if (numBuffers > INT_MAX) {
+        JNI_TRACE("too many buffers");
+        Errors::jniThrowRuntimeException(env, "too many buffers");
+        return nullptr;
+    }
+
+    ScopedLocalRef<jobjectArray> array(
+            env, env->NewObjectArray(static_cast<int>(numBuffers), JniConstants::byteArrayClass,
+                                     nullptr));
+    if (array.get() == nullptr) {
+        JNI_TRACE("failed to allocate array");
+        return nullptr;
+    }
+
+    for (size_t i = 0; i < numBuffers; ++i) {
+        CRYPTO_BUFFER* buffer = sk_CRYPTO_BUFFER_value(buffers, i);
+        ScopedLocalRef<jbyteArray> bArray(env, CryptoBufferToByteArray(env, buffer));
+        if (bArray.get() == nullptr) {
+            return nullptr;
+        }
+        env->SetObjectArrayElement(array.get(), i, bArray.get());
```

Comment:
The code is correct as-is w.r.t. to the ScopedLocalRef.

Every loop iteration it will destroy the ScopedLocalRef which calls JNIEnv::DeleteLocalRef. If you call "release" instead of "get" on L409, it will retain local ref for every single loop iteration.

It "numBuffers" is too large, JNI will eventually exhaust local refs and subsequent JNI calls that need new refs will fail.

----
As for JNI local ref ownership, it's quite simple. Any JNI function that returns a jobject back to native code will have an implicit local ref. 

A local ref will act as a GC root, preventing that object from being garbage collected. Local refs get cleaned up automatically in one of two ways:

1) When you *return* from JNI back to Java code (i.e. the stack frames all unwind and you end up in Java code again): all the local refs are deleted. 

2) Calling PopLocalFrame: cleans up all local refs that were made since the last PushLocalFrame.

--------
If memory serves we have nearly unlimited local refs starting with Android-O provided that PushLocalFrame is used every once in a while; EnsureLocalCapacity can be used to detect an appropriate size.